pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/AvoidInstantiatingObjectsInLoopsRule.java

```diff
@@ -14,24 +14,77 @@
 import net.sourceforge.pmd.lang.java.ast.ASTWhileStatement;
 
 public class AvoidInstantiatingObjectsInLoopsRule extends AbstractOptimizationRule {
-
+    /**
+     * This method is used to check whether user instantiates variables
+     * which are not assigned in loops.
+     * @param node This is the expression of part of java code to be checked.
+     * @param data This is the data to return.
+     * @return Object This returns the data passed in. If violation happens, violation is added to data.
+     */
     @Override
     public Object visit(ASTAllocationExpression node, Object data) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
         if (insideLoop(node) && fourthParentNotThrow(node) && fourthParentNotReturn(node)) {
-            addViolation(data, node);
+            if (thirdParentNotASTExpression(node) && fourthParentNotASTStatementExpression(node)) {
+                addViolation(data, node);
+            }
         }
         return data;
     }
 
-    private boolean fourthParentNotThrow(ASTAllocationExpression node) {
+    /**
+     * This method is used to check whether the instantiated variable is assigned or not.
+     * @param node This is the expression of part of java code to be checked.
+     * @return boolean This returns Whether the third parent of node is an ASTExpression.
+     */
+    public boolean thirdParentNotASTExpression(ASTAllocationExpression node) {
+        //CS304 Issue link: https://github.com/pmd/pmd/issues/2207
+        if (node.getParent().getClass().toString().equals(
+                "class net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix")
```

Comment:
Please don't use the `toString()` method for any logic. The `toString()` method is there for debugging purposes and in most cases the format is not an API and can change without any notice. Relying on the format of `toString()` makes java programs brittle. Also, in that case, if we decided to refactor the classes and rename/move ASTPrimaryPrefix, then we won't get here a compile time error but only a runtime error during execution - which is too late.

What you want to achieve here is, to test, whether the parent (or grandparent or greatgrandparent) node is a instance of a specific type:

```java
if (node.getParent() instanceof ASTPrimaryPrefix
    && node.getNthParent(2) instanceof ASTPrimaryExpression) {
    return !(node.getNthParent(3) instanceof ASTExpression);
}
return false;
```